From f36fe7ab6e8471cdc344fa8d9430b901dae7b93e Mon Sep 17 00:00:00 2001
From: Andrew Vit <andrew@avit.ca>
Date: Sun, 15 Feb 2009 19:33:12 -0800
Subject: [PATCH] Added :env option for gem in template runner

For installing gems that are only needed in the test environment, specify the :env option so the dependency is written to config/environments/test.rb:

    gem 'rspec', :env => 'test'
    gem 'quietbacktrace', :env => %w[development test]
---
 .../generators/applications/app/template_runner.rb |   28 +++++++++++++++++---
 .../test/generators/rails_template_runner_test.rb  |   11 ++++++++
 2 files changed, 35 insertions(+), 4 deletions(-)

diff --git a/railties/lib/rails_generator/generators/applications/app/template_runner.rb b/railties/lib/rails_generator/generators/applications/app/template_runner.rb
index 84e36ec..e55e271 100644
--- a/railties/lib/rails_generator/generators/applications/app/template_runner.rb
+++ b/railties/lib/rails_generator/generators/applications/app/template_runner.rb
@@ -85,6 +85,7 @@ module Rails
     # Adds an entry into config/environment.rb for the supplied gem :
     def gem(name, options = {})
       log 'gem', name
+      env = options.delete(:env)
 
       gems_code = "config.gem '#{name}'"
 
@@ -93,18 +94,26 @@ module Rails
         gems_code << ", #{opts}"
       end
 
-      environment gems_code
+      environment gems_code, :env => env
     end
 
     # Adds a line inside the Initializer block for config/environment.rb. Used by #gem
-    def environment(data = nil, &block)
+    # If options :env is specified, the line is appended to the corresponding
+    # file in config/environments/#{env}.rb
+    def environment(data = nil, options = {}, &block)
       sentinel = 'Rails::Initializer.run do |config|'
 
       data = block.call if !data && block_given?
 
       in_root do
-        gsub_file 'config/environment.rb', /(#{Regexp.escape(sentinel)})/mi do |match|
-          "#{match}\n " << data
+        if options[:env].nil?
+          gsub_file 'config/environment.rb', /(#{Regexp.escape(sentinel)})/mi do |match|
+            "#{match}\n  " << data
+          end
+        else
+          options[:env].to_a.each do|env|
+            append_file "config/environments/#{env}.rb", "\n#{data}"
+          end
         end
       end
     end
@@ -350,6 +359,17 @@ module Rails
       File.open(path, 'wb') { |file| file.write(content) }
     end
 
+    # Append text to a file
+    #
+    # ==== Example
+    #
+    #   append_file 'config/environments/test.rb', 'config.gem "rspec"'
+    #
+    def append_file(relative_destination, data)
+      path = destination_path(relative_destination)
+      File.open(path, 'ab') { |file| file.write(data) }
+    end
+
     def destination_path(relative_destination)
       File.join(root, relative_destination)
     end
diff --git a/railties/test/generators/rails_template_runner_test.rb b/railties/test/generators/rails_template_runner_test.rb
index fcc0206..fd14822 100644
--- a/railties/test/generators/rails_template_runner_test.rb
+++ b/railties/test/generators/rails_template_runner_test.rb
@@ -82,6 +82,17 @@ class RailsTemplateRunnerTest < GeneratorTestCase
     assert_rails_initializer_includes("config.gem 'mislav-will-paginate', :lib => 'will-paginate', :source => 'http://gems.github.com'")
   end
 
+  def test_gem_with_env_string_should_put_gem_dependency_in_specified_environment
+    run_template_method(:gem, 'rspec', :env => 'test')
+    assert_generated_file_with_data('config/environments/test.rb', "config.gem 'rspec'", 'test')
+  end
+
+  def test_gem_with_env_array_should_put_gem_dependency_in_specified_environments
+    run_template_method(:gem, 'quietbacktrace', :env => %w[ development test ])
+    assert_generated_file_with_data('config/environments/development.rb', "config.gem 'quietbacktrace'")
+    assert_generated_file_with_data('config/environments/test.rb', "config.gem 'quietbacktrace'")
+  end
+
   def test_environment_should_include_data_in_environment_initializer_block
     load_paths = 'config.load_paths += %w["#{RAILS_ROOT}/app/extras"]'
     run_template_method(:environment, load_paths)
-- 
1.6.1.2

